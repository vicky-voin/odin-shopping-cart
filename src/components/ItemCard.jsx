import useProductData from "../hooks/useProductData";

export default function ItemCard({ itemId }) {
  const { data, isError } = useProductData(itemId);

  return (
    <div>
      {data !== null && (
        <div>
          <div>{data.title}</div>
          <div>{data.description}</div>
          <img src={data.imageUrl}></img>
        </div>
      )}
      {isError && <div>Error: could not load item</div>}
    </div>
  );
}
