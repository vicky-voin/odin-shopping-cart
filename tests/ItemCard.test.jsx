import { vi, describe, it, expect, afterEach } from "vitest";
import { screen, render } from "@testing-library/react";
import ItemCard from "../src/components/ItemCard";
import formatInUSD from "../src/utils/priceConverter";

describe("Item Card component tests", () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  it("renders a skekelton card when no data provided", () => {
    const { container } = render(<ItemCard></ItemCard>);
    expect(container).toMatchSnapshot();
  });

  it("fills out the item fields correctly", async () => {
    const data = getFakeData();
    globalThis.fetch = vi.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => data,
      })
    );

    render(<ItemCard></ItemCard>);

    const formattedPrice = formatInUSD(data.price);

    await screen.findByText(data.title);
    await screen.findByText(data.description);
    await screen.findByText(formattedPrice);
    const img = await screen.findByRole("img");
    expect(img).toHaveAttribute("src", data.imageUrl);
  });

  it("renders error message in card on error", async () => {
    globalThis.fetch = vi.fn(() =>
      Promise.resolve({
        ok: false,
        json: () => "Server error: could not fetch item",
      })
    );

    render(<ItemCard></ItemCard>);

    await screen.findByText("Error: could not load item");
  });

  function getFakeData() {
    return {
      title: "Fake Item",
      description: "Fake Item description",
      image: "www.google.com",
      price: "120.5",
    };
  }
});
