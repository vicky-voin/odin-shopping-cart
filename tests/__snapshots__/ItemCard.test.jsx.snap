// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Item Card component tests > renders a default card 1`] = `
<div>
  <div />
</div>
`;

exports[`Item Card component tests > renders an empty card when no data provided 1`] = `
<div>
  <div />
</div>
`;
